File: processing/syntactic/expressions/unknown_expr.py
from __future__ import annotations

from errors.blame import BlameType
from processing.lexical.tokens.token import Token
from processing.lexical.tokens.token_type import TokenType
from processing.syntactic.expressions.expr import Expr
from processing.location import span_marker


class UnknownExpr(Expr):
    """ expr:
        TOKEN* (NEWLINE | END);
    """

    def __init__(
            self,
            parent: Expr = None,
            *tokens: Token
    ):
        super().__init__(parent)
        self.tokens = list(tokens)
        self.source.blame(BlameType.invalid_syntax, self)

    @span_marker
    def parse(self) -> UnknownExpr:
        while not self.stream.exact_peek.of_type(TokenType.newline, TokenType.end):
            self.tokens.append(self.stream.eat_any())
        return self

File: Axion.Core/Processing/Syntactic/Expressions/UnknownExpr.cs
using Axion.Core.Processing.Errors;
using Axion.Core.Processing.Lexical.Tokens;
using Axion.Core.Processing.Syntactic.Expressions.Common;
using Axion.Specification;
using Magnolia.Attributes;
using Magnolia.Trees;
using static Axion.Specification.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions;

/// <summary>
///     <code>
///         unknown-expr:
///             TOKEN* (NEWLINE | END);
///     </code>
/// </summary>
[Branch]
public partial class UnknownExpr : AtomExpr {
    [Leaf] NodeList<Token, Ast>? tokens;

    public UnknownExpr(Node parent) : base(parent) { }

    public UnknownExpr Parse() {
        while (!Stream.PeekIs(Newline, TokenType.End)) {
            Tokens += Stream.Eat();
        }
        LanguageReport.To(BlameType.InvalidSyntax, this);
        return this;
    }
}
